File: leetcode-new/211.cpp
class TrieNode
{
public:
    unordered_map<char, TrieNode *> children;
    bool isEnd;
    TrieNode() : children{}, isEnd{false} {}
};

class WordDictionary
{
private:
    TrieNode *head;

public:
    WordDictionary()
    {
        head = new TrieNode();
    }

    void addWord(string word)
    {
        TrieNode *curr = head;
        for (char c : word)
        {
            auto &wl = curr->children;
            if (wl.find(c) == wl.end())
                wl[c] = new TrieNode();
            curr = wl[c];
        }
        curr->isEnd = true;
    }

    bool search(string word)
    {
        return helper(word, 0, head);
    }

private:
    bool helper(string &word, int idx, TrieNode *curr)
    {
        if (idx == word.length())
            return curr->isEnd;
        char c = word[idx];
        auto &wl = curr->children;
        if (c == '.')
        {
            for (auto &pair : wl)
            {
                if (helper(word, idx + 1, pair.second))
                    return true;
            }
            return false;
        }
        else
        {
            if (wl.find(c) == wl.end())
                return false;
            return helper(word, idx + 1, wl[c]);
        }
    }
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: leetcode-larry/211.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.isWord = False

class WordDictionary:
    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        curr = self.root

        for i in word:
            if i not in curr.children:
                curr.children[i] = TrieNode()
            curr = curr.children[i]

        curr.isWord = True

    def search(self, word: str) -> bool:
        def helper(index, curr):
            for i in range(index, len(word)):
                if word[i] == ".":
                    for v in curr.children.values():
                        if helper(i + 1, v):
                            return True
                    return False
                else:
                    if word[i] not in curr.children:
                        return False
                    curr = curr.children[word[i]]

            return curr.isWord

        return helper(0, self.root)

# Your WordDictionary object will be instantiated and called as such:
# obj = WordDictionary()
# obj.addWord(word)
# param_2 = obj.search(word)